Add API tests for skipping ads in PIP

danae404 opened the pull request:
#### 5b4ed0533cea60126d97815660295db151ae86f0
<pre>
Add API tests for skipping ads in PIP
<a href="https://bugs.webkit.org/show_bug.cgi?id=289744">https://bugs.webkit.org/show_bug.cgi?id=289744</a>
<a href="https://rdar.apple.com/146994962">rdar://146994962</a>

Reviewed by Jer Noble.

This patch adds two API tests for ad skipping in PIP. The first
test checks that we set the PIP playbackstate&apos;s PrerollAttributes
property to a nonnull  value when the webpage has set the media session
skipad action handler and a video is in pip.

The second test checks the prerollAttributes stays null when a video
goes into pip and the webpage has not set the skipAd action handler.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.html: Added.
* Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm: Added.
(-[SkipAdInPiPFullscreenDelegate _webView:hasVideoInPictureInPictureDidChange:]):
(TestWebKitAPI::createWebView):
(TestWebKitAPI::TEST(SkipAdInPictureInPicture, VideoHasAd)):
(TestWebKitAPI::TEST(SkipAdInPictureInPicture, VideoHasNoAd)):

Canonical link: <a href="https://commits.webkit.org/302095@main">https://commits.webkit.org/302095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36d592edc916523e3f3b1bc0842826bfb26a61ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135387 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79528 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6e655749-22c7-48ff-a889-3257d18b8e44) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129888 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/170 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/97452 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/11157530-e69b-4e88-aa29-7d7213801a78) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130964 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114679 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78021 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/18ee910d-2015-4e19-998c-b4cf91380d42) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32784 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78696 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/108472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33264 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137870 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/155 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/145 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105979 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105716 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/117 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29578 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52323 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20001 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/201 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/120 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/197 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/161 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -6644,6 +6651,7 @@
 					};
 					A17C46452C98E3430023F3C7 = {
 						CreatedOnToolsVersion = 16.0;
+						LastSwiftMigration = 1700;
```
Please revert this change.

pxlcoder commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -7924,13 +7933,18 @@
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
+				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_VERSION = 6.0;
```
Ditto.

pxlcoder commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -7924,13 +7933,18 @@
 			baseConfigurationReference = A17C464A2C98E3640023F3C7 /* TestWebKitAPIResources.xcconfig */;
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
+				SWIFT_VERSION = 6.0;
```
Ditto.

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm:
```diff
@@ -0,0 +1,144 @@
+SOFT_LINK_CLASS(PIP, PIPPrerollAttributes)
+SOFT_LINK_CLASS(PIP, PIPViewController)
+
+static bool didEnterPiP;
```
```suggestion
static bool didEnterPiP = false;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm:
```diff
@@ -0,0 +1,144 @@
+    ASSERT_TRUE(pipPanel);
+    auto *contentViewController = [pipPanel contentViewController];
+    ASSERT_TRUE([contentViewController isKindOfClass:getPIPViewControllerClass()]);
+    auto *pipViewController = (PIPViewController*)contentViewController;
```
```suggestion
    auto *pipViewController = (PIPViewController *)contentViewController;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm:
```diff
@@ -0,0 +1,144 @@
+
+TEST(SkipAdInPictureInPicture, VideoHasAd)
+{
+    auto webView = createWebView();
```
Prefer use of explicit `RetainPtr` over `auto`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm:
```diff
@@ -0,0 +1,144 @@
+        }
+    }];
+    ASSERT_TRUE(pipPanel);
+    auto *contentViewController = [pipPanel contentViewController];
```
Not the most important for testing code, but it's a good habit to start using `RetainPtr` everywhere instead of raw pointers.

rr-codes commented on Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm:
```diff
@@ -0,0 +1,144 @@
+    auto *pipViewController = (PIPViewController*)contentViewController;
+    ASSERT_TRUE(pipViewController);
+
+    auto *pipPlaybackState = pipViewController.playbackState;
```
```suggestion
    RetainPtr pipPlaybackState = pipViewController.playbackState;
```

rr-codes commented on Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm:
```diff
@@ -0,0 +1,144 @@
+        }
+    }];
+    ASSERT_TRUE(pipPanel);
+    auto *contentViewController = [pipPanel contentViewController];
```
ditto here and below

danae404 committed:
Add API tests for skipping ads in PIP
https://bugs.webkit.org/show_bug.cgi?id=289744
rdar://146994962

Reviewed by Jer Noble.

This patch adds two API tests for ad skipping in PIP. The first
test checks that we set the PIP playbackstate's PrerollAttributes
property to a nonnull  value when the webpage has set the media session
skipad action handler and a video is in pip.

The second test checks the prerollAttributes stays null when a video
goes into pip and the webpage has not set the skipAd action handler.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.html: Added.
* Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm: Added.
(-[SkipAdInPiPFullscreenDelegate _webView:hasVideoInPictureInPictureDidChange:]):
(TestWebKitAPI::createWebView):
(TestWebKitAPI::TEST(SkipAdInPictureInPicture, VideoHasAd)):
(TestWebKitAPI::TEST(SkipAdInPictureInPicture, VideoHasNoAd)):

Canonical link: https://commits.webkit.org/302095@main
